### DISCUSSION TITLE
Cloud detection

### DISCUSSION THREAD

**Discussion by evripidis** [aaronwmorris/indi-allsky#951] [Ideas]
Just an idea: If there is any way that could make some kind of calculations so we would be able to use it as a some kind of cloud detector.

Going further, the idea is to "integrate" somehow the project with KStars, so we could use this ability while we are capturing photos.

**Comment by aaronwmorris** [comment]
There was a very productive post over on CloudyNights regarding this topic.  It focused on using Machine Learning to classify images:

https://www.cloudynights.com/topic/823536-cloud-detection-for-all-sky-cameras/

Perhaps when I get my module system off the ground, this can be provided as a plugin, but this is really outside the scope of core features for an all sky camera system.

**Reply by evripidis** [reply]
That seems pretty interesting yeah, thank you!

Of course, this is an extra and even an advanced feature but besides my thoughts for integration with KStars, that kind of detection it could be really handy for other occasions too, for example clouds -> rain is coming so I have to take care something outside in the garden. 

"If clouds detected run this script.sh that will text me" etc.

Just thinking out loud :D

**Reply by aaronwmorris** [reply]
The star count is probably one of the best ways to interpret the cloud situation.  If you have light pollution, the SQM value will usually increase if it gets cloudy.  Most of the data can be piped to a MQTT broker.  A few people have setup alerts based on the MQTT data.

**Reply by evripidis** [reply]
Yes probably this would be an eligible way to have something but first I have to make more adjustments on that, not only the mask. My location (city) is not the best and yes those 22 are certainly not stars. ![Screenshot 2023-09-21 at 20 45 24](https://github.com/aaronwmorris/indi-allsky/assets/216827/983f8cfd-f750-47b8-815e-fc1e362d3291)

**Comment by caphector** [comment]
I've been working on cloud recognition with an AI model; still tinkering with it, but I can pass it an image and have it tell me what clouds it detects. Still working out stuff like running on a Pi directly; thought it would be interesting to record cloud types as they go overhead.

**Reply by Mooflotic** [reply]
Me as well, I was tinkering with https://github.com/gordtulloch/mlCloudDetect

**Reply by caphector** [reply]
Mine tries to classify clouds into the various cloud types as well; all I've done is train a model. Have been hitting dependency hell trying to get it converted to a model that can run on Pi.

**Reply by caphector** [reply]
Got it working on my server on my "current image" page. The classifier is run via `cron` at the moment and outputs to a text file, which is then read and displayed by the page.

https://caphector.com/allsky/ (shows in the sidebar as something like `Cirrocumulus clouds (88.69% confidence)`)

Unrelated; the calendar view is semi-broke, but I wanted to get *something* to show the cloud recognizer working. Haven't tested it on night-sky images yet. I don't use the indi-allsky webinterface because it uses a lot of space.

**Reply by Vertikar** [reply]
> Got it working on my server on my "current image" page. The classifier is run via `cron` at the moment and outputs to a text file, which is then read and displayed by the page.

That's really cool! So does your server have the image transferred to it from indi-allsky, then cron periodically kicks off a script to classify the most recent image?

**Reply by caphector** [reply]
Yep; I use the indi-allsky uploader and then kick off the script via Cron once a minute (with some logic to check for a reasonable file size to avoid scanning a partial upload). Each scan takes ~5 seconds of system time.
